File: python-algorithm/ds_4_linkedlist/linkedlist_3_two_points/L83.cpp
//
// Created by LiHeng on 2020/7/19.
//
#include <iostream>
using namespace std;

//  Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

//非递归写法，设置一个标记上一次重复值第一次出现的节点
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if(head == NULL || head->next == NULL) return head;
        ListNode* start = head;
        ListNode* pre = head;//标记上个相同值
        ListNode* cur = head->next;
        while(cur){
            if(cur->val != pre->val){
                pre->next = cur;
                pre = cur;
            }
            cur = cur->next;
        }
        pre->next = NULL;
        return start;
    }
};

//递归写法，有点不好理解
class Solution1 {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (head == NULL || head->next == NULL) return head;
        head->next = deleteDuplicates(head->next);
        //cout <<head->val<< head->next->val<<endl;
        //return的东西就是上层递归的返回
        return head->val == head->next->val ? head->next : head;
    }
};

File: python-algorithm/ds_4_linkedlist/linkedlist_3_two_points/L83.py
#--coding:utf-8--

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if not head:
            return None
        back = head
        cur = head.next
        while cur:
            if cur.val == back.val:
                cur = cur.next
            else:
                back.next = cur
                back = cur
                cur = cur.next
        # 这句容易漏掉，处理最后是多个重复的数字
        back.next = None
        return head


class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        flag = head.val
        back = head
        cur = head.next
        while cur :
            if cur.val == flag:
                back.next = cur.next
                cur = cur.next
            else:
                back = cur
                flag = back.val
                cur = cur.next
        return head